crates/codegen/syntax_templates/src/typescript/parse_output.rs

```diff
@@ -0,0 +1,64 @@
+use std::{collections::BTreeSet, ops::Range};
+
+use super::{
+    cst, cst_types::RcNodeExtensions as CSTRcNodeExtensions, kinds::TokenKind,
+    parse_error::render_error_report,
+};
+use napi::bindgen_prelude::*;
+
+#[napi]
+pub struct ParseOutput {
+    pub(crate) parse_tree: cst::Node,
+    pub(crate) errors: Vec<ParseError>,
+}
+
+#[napi]
+impl ParseOutput {
+    #[napi(getter, ts_return_type = "RuleNode | TokenNode | null")]
+    pub fn parse_tree(&self, env: Env) -> napi::JsObject {
+        return self.parse_tree.clone().to_js(&env);
+    }
+
+    #[napi(getter)]
+    pub fn errors(&self) -> Vec<ParseError> {
+        return self.errors.clone();
+    }
+
+    #[napi(getter)]
+    pub fn is_valid(&self) -> bool {
+        return self.errors.is_empty();
+    }
+}
+
+#[napi]
+#[derive(PartialEq, Clone)]
+pub struct ParseError {
+    pub(crate) utf8_range: Range<usize>,
+    pub(crate) expected_tokens: Vec<TokenKind>,
+}
+
+#[napi]
+impl ParseError {
+    #[napi(getter, ts_return_type = "[ start: number, end: number ]")]
+    pub fn utf8_range(&self) -> [u32; 2] {
```

Comment:
Same question here, about using `Range<TextIndex` and exposing the other types.

Reply:
done